src/audio/dcblock/dcblock_hifi4.c

```diff
@@ -0,0 +1,172 @@
+LOG_MODULE_DECLARE(dcblock, CONFIG_SOF_LOG_LEVEL);
+
+static inline ae_int32x2  dcblock_cal(ae_int32x2 R, ae_int32x2 state_x, ae_int32x2 state_y,
+				      ae_int32x2 sample)
+{
+	ae_int64 out, temp;
+
+	/* R: Q2.30, y_prev: Q1.31 the result is Q2.62 */
+	temp = AE_MULF32S_LL(R, state_y);
+	/* shift temp to 1.63 */
+	temp = AE_SLAI64S(temp, 1);
+
+	out = AE_SUB64S(AE_CVT64F32_L(sample), AE_CVT64F32_L(state_x));
+	out = AE_ADD64S(out, temp);
+
+	return AE_ROUND32F64SSYM(out);
+}
+
+/* Setup circular for component sink and source */
+static inline void dcblock_set_circular(const struct audio_stream __sparse_cache *source,
+					const struct audio_stream __sparse_cache *sink)
+{
+	/* Set source as circular buffer 0 */
+	AE_SETCBEGIN0(source->addr);
+	AE_SETCEND0(source->end_addr);
+
+	/* Set sink as circular buffer 1 */
+	AE_SETCBEGIN1(sink->addr);
+	AE_SETCEND1(sink->end_addr);
+}
+
+#if CONFIG_FORMAT_S16LE
+static void dcblock_s16_default(const struct comp_dev *dev,
+				const struct audio_stream __sparse_cache *source,
+				const struct audio_stream __sparse_cache *sink,
+				uint32_t frames)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	ae_int16 *in;
+	ae_int16 *out;
+	ae_int32x2 R, state_x, state_y, sample;
+	ae_int16x4 in_sample, out_sample;
+	int ch, i;
+	int nch = source->channels;
+	const int inc = nch * sizeof(ae_int16);
+
+	dcblock_set_circular(source, sink);
+	for (ch = 0; ch < nch; ch++) {
+		in = (ae_int16 *)source->r_ptr + ch;
+		out = (ae_int16 *)sink->w_ptr + ch;
+		state_x = cd->state[ch].x_prev;
+		state_y = cd->state[ch].y_prev;
+		R = cd->R_coeffs[ch];
+		for (i = 0; i < frames; i++) {
+			/* Load a 16 bit sample*/
+			AE_L16_XC(in_sample, in, inc);
+			/* store the 16 bit sample to high 16bit of 32bit register*/
+			sample = AE_CVT32X2F16_32(in_sample);
+			state_y = dcblock_cal(R, state_x, state_y, sample);
+			state_x = sample;
+			out_sample = AE_ROUND16X4F32SSYM(state_y, state_y);
+			AE_S16_0_XC1(out_sample, out, inc);
+		}
+		cd->state[ch].x_prev = state_x;
+		cd->state[ch].y_prev = state_y;
+	}
+}
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE
+static void dcblock_s24_default(const struct comp_dev *dev,
+				const struct audio_stream __sparse_cache *source,
+				const struct audio_stream __sparse_cache *sink,
+				uint32_t frames)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	ae_int32 *in;
+	ae_int32 *out;
+	ae_int32x2 R, state_x, state_y;
+	ae_int32x2 in_sample, out_sample;
+	int ch, i;
+	int nch = source->channels;
+	const int inc = nch * sizeof(ae_int32);
+
+	dcblock_set_circular(source, sink);
+	for (ch = 0; ch < nch; ch++) {
+		in = (ae_int32 *)source->r_ptr + ch;
+		out = (ae_int32 *)sink->w_ptr + ch;
+
+		state_x = cd->state[ch].x_prev;
+		state_y = cd->state[ch].y_prev;
+		R = cd->R_coeffs[ch];
+		for (i = 0; i < frames; i++) {
+			AE_L32_XC(in_sample, in, inc);
+			in_sample = AE_SLAI32(in_sample, 8);
+			state_y = dcblock_cal(R, state_x, state_y, in_sample);
```

Comment:
except for 2 circular buffers support in hifi4,  another important thing is: it support 4 32 * 32 per cycle, this means, if possible, there should have 4 multiply in parallel, take 1ms 48k as example:
for loop may be like below:
for (i = 0; i < frames/4; i++) {};

if have 47/49 case, we need handle it separately.

Reply:
I'm afraid you have to know the difference between "frames" and "samples" :).  In our SOF, a frame contains channel_num samples. In this component, we processed data by channel.